### PR TITLE
Add assertions to enforce only one dirty bit set at a time in Halide::Runtime::Buffer.

### PR DESCRIPTION
**Pull request by zvookin**
on entry. This prevents a common bug where both dirty bits end up
being set. (And prevents reading stale data when a buffer has been
access on the device last.) An alternative design would be to
move dirty bit management out to a higher, explicitly written,
level. The current design feels like the worst of both worlds in that
it automatically manages half the state but not the other half.

Removed a memset to zero of the buf field which is redundant due to
the default initializer on buf.

### CODE REVIEW

**Comment by zvookin** [comment]
(This is related to the bug Ron reported in HelloHexagon after adding the device interface asserts. That code doesn't fail correctness because the buffers are zero copy.)

**Comment by abadams** [comment]
I worry that this will destroy the performance of code that uses operator() - even if it gets inlined, the conditional branch in the copy_to_host may prevent autovectorization of those loops.

**Comment by zvookin** [comment]
Yeah, I opened the PR because this is clearly a performance issue. I don't think I can even put in an assert that the device is not dirty without hitting the same problem. Dillon and I discussed it and decided if one wants performance, writing the loop in Halide is the way to go. (The last is not meant to be snark. If this particular problem had a good solution, template metaprogramming would be much more viable for this kind of thing.)

I'm happy to go the other way and remove the setting of host dirty out of the operator, which is probably the consistent design. E.g. If we're handling this automatically, for_each_value with non-const refs should set host dirty. (And for_each_value always ought to call copy_to_host before it starts accessing values.)

Another design would be to add scoped access methods that manage the transitions. One could either do this via convention or enforce it statically by having the operator() be on a helper class. We could even provide both forms of access depending on what level or performance vs. convenience is desired. (And keeping backward compatibility.)

(To make the latter more concrete, I'm thinking:

```
    class BufferView {
        enum class Access { Read, Write };
        BufferView(Buffer &buf, Access access);

        operator(...) { } // indexing operator.
    }

```
One can have a constructor that takes a const Buffer reference which implies read access. One could also pass in a cropping rectangle for iterator patterns, etc. Tiled memory designs generaly end up looking like this as the outer object doesn't have everything in memory.)

**Comment by ronlieb** [comment]
FYI, after Dillon merged the change "Copy output back to the host explicitly."
the error went away ...

File: src/runtime/HalideBuffer.h [hunk 1]
```diff
@@ -474,7 +474,6 @@ class Buffer {
     }
 
     Buffer() {
-        memset(&buf, 0, sizeof(halide_buffer_t));
```
Comment (steven-johnson) [inline src/runtime/HalideBuffer.h]: Why was this removed?

**Comment by zvookin** [comment]
The call to copy_to_host has to happen, though on zero copy setups one might be able to get away without it. The issue is the behavior of operator() makes it too easy to get the dirty bits in an inconsistent state if one forgets the copy_to_host. (device_interface now signals that error, but that would be quite a bit later, if it hits at all.)

I'm advocating with either automate it, so the copy_to_host just happens, or mandate that all dirty bit management happen explicitly.

File: src/runtime/HalideBuffer.h (hunk 1 above)
Comment (zvookin) [inline src/runtime/HalideBuffer.h]: Because the default member initializer does it and we weren't doing it in other constructors so the code confused me.

**Comment by zvookin** [comment]
Changed direction here per conversation with Andrew.

File: apps/support/Makefile.inc [hunk 2]
```diff
@@ -55,9 +55,16 @@ endif
 endif
 LIBPNG_LIBS ?= $(LIBPNG_LIBS_DEFAULT)
 
-LIBJPEG_LIBS = -ljpeg
+# Workaround brew Cellar path for libpng-config output.
+LIBJPEG_LINKER_PATH ?= $(shell echo $(LIBPNG_LIBS_DEFAULT) | sed -e'/-L.*[/][Cc]ellar[/]libpng/!d;s=\(.*\)/[Cc]ellar/libpng/.*=\1/lib=')
```
Comment (abadams) [inline apps/support/Makefile.inc]: What's going on here?

Comment (zvookin) [reply apps/support/Makefile.inc]: This was done to the toplevel Makefile in a PR a couple weeks or so ago. Brew installs each package in its own directory and then links to the commands from the install directory. The -config command typically returns the per-package directory. This heuristically recognizes that situation and makes a better guess for the case. It means I can build on the Mac and shouldn't affect anything else.

We should probably have a single Makefile include that contains functionality used everywhere, including the top-level. I was hoping we'd be using bazel before we had to start doing stuff like this.

**Comment by abadams** [comment]
lgtm. I merged master in to make sure the windows failures were due to that other issue
